### ISSUE TITLE
Test failure (api/cpp/sep_log_api) on 32-bit architectures

### ISSUE THREAD

**Issue by swt2c** [cvc5/cvc5#9366]
I'm seeing the api/cpp/sep_log_api test fail on 32-bit architectures (i386, armel, armhf).  This is on Debian Sid with cvc5 1.0.3.

```
3165/3185 Test #3225: api/cpp/reset_assertions .....................................................................   Passed    0.01 sec
          Start 3226: api/cpp/sep_log_api
3166/3185 Test #3226: api/cpp/sep_log_api ..........................................................................Subprocess aborted***Exception:   0.02 sec
terminate called after throwing an instance of 'cvc5::CVC5ApiException'
  what():  Invalid argument '0' for '*d_node', expected Term to be a 64-bit integer value when calling getInt64Value()

          Start 3227: api/cpp/smt2_compliance
3167/3185 Test #3227: api/cpp/smt2_compliance ......................................................................   Passed    0.01 sec
```

Any advice on how to investigate further would be appreciated.  CC @nafur since you fixed a similar issue with this test?

**Comment by mpreiner** [comment]
Hi @swt2c, thanks for reporting! These platforms are currently not supported but we are trying to get better support in the future. The i386 one is interesting since we have i386 Ubuntu nightly builds that don't fail. We'll investigate.

**Comment by mpreiner** [comment]
Yes, we still have i386 Debian nightlies and they pass:

https://cvc5-buildbot.stanford.edu/#/builders/36

**Comment by swt2c** [comment]
> Yes, we still have i386 Debian nightlies and they pass:
> 
> https://cvc5-buildbot.stanford.edu/#/builders/36

Is the buildbot worker/job configuration publicly viewable anywhere so that I might try to spot the difference(s)?

**Comment by mpreiner** [comment]
You can find the cvc5 configuration here: https://cvc5-buildbot.stanford.edu/#/builders/36/builds/125/steps/2/logs/stdio

Here's the Dockerfile we use for this worker:

```
ARG ARCH=
ARG VERSION=
FROM buildbot-worker:debian${VERSION}-${ARCH}

ARG DEBIAN_FRONTEND=noninteractive

# Setup cvc5 specific dependencies
USER root
RUN apt-get update && apt-get install -y \
  autoconf \
  automake \
  build-essential \
  clang \
  cmake \
  curl \
  default-jdk \
  default-jre \
  flex \
  gcovr \
  lcov \
  libcln-dev \
  libgmp-dev \
  libgtest-dev \
  libreadline-dev \
  libtool \
  mingw-w64 \
  ninja-build \
  pkg-config \
  python3 \
  python3-pip \
  wget \
  zip
RUN pip3 install toml
RUN pip3 install scikit-build
RUN pip3 install pytest
RUN pip3 install pyparsing
RUN pip3 install cython==0.29  --install-option="--no-cython-compile"
RUN pip3 install fastcov
RUN cd /usr/src/gtest && cmake . && make && cp $(find . -name "libgtest*.a") /lib/

# ANTLR3 configure does not find types.h in /usr/include/aarch64-linux-gnu/sys/
RUN if [ "$(uname -m)" = "aarch64" ]; then ln -s /usr/include/aarch64-linux-gnu/sys/ /usr/include/sys; fi

# cvc5 requires the POSIX version of MinGW to implement the time limit using
# threads
RUN update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-posix
RUN update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix

ENV PYTHONIOENCODING=utf-8
USER buildbot
```

**Comment by swt2c** [comment]
Do you know where the `buildbot-worker:debian` is coming from?  Because if it's from here:
https://hub.docker.com/r/buildbot/buildbot-worker

I'm a bit confused as that image seems to be based on Ubuntu, as best as I can tell.

**Comment by mpreiner** [comment]
We modified the Ubuntu worker image for Debian:

```
ARG ARCH=
ARG VERSION=
FROM ${ARCH}debian:${VERSION}

ARG ARCH
ARG DEBIAN_FRONTEND=noninteractive

# Install security updates and required packages
RUN apt-get update && \
    apt-get -y upgrade && \
    apt-get -y install -q \
    build-essential \
    git \
    python-dev \
    libffi-dev \
    libssl-dev \
    python3-setuptools \
    python3-pip \
    # Test runs produce a great quantity of dead grandchild processes.  In a
    # non-docker environment, these are automatically reaped by init (process 1),
    # so we need to simulate that here.  See https://github.com/Yelp/dumb-init
    dumb-init \
    curl \
    cargo

RUN rm -rf /var/lib/apt/lists/*

# Install required python packages, and twisted
RUN python3 -m pip install -U pip
RUN pip3 --no-cache-dir install 'twisted[tls]'
RUN pip3 install virtualenv
RUN pip3 install buildbot-worker

RUN useradd -ms /bin/bash buildbot

RUN mkdir /buildbot
COPY buildbot.tac /buildbot/buildbot.tac
COPY init-buildbot-worker.sh /init-buildbot-worker.sh
RUN chmod +x /init-buildbot-worker.sh
RUN chown -R buildbot /buildbot

USER buildbot
WORKDIR /buildbot

CMD ["/init-buildbot-worker.sh"]
```

**Comment by mpreiner** [comment]
@swt2c Do these issues still occur on your side?

**Comment by swt2c** [comment]
> @swt2c Do these issues still occur on your side?

Yes, they do.  :(

**Comment by mpreiner** [comment]
Is there maybe a docker-based environment to reproduce the issue? Our Debian i386 builds don't fail.

**Comment by swt2c** [comment]
I will try to create one.

**Comment by aytey** [comment]
I can reproduce this with:

```
uname -a
Linux deb-12-4 6.1.0-17-686-pae #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) i686 GNU/Linux

lsb_release  -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm

ldd --version
ldd (Debian GLIBC 2.36-9+deb12u4) 2.36
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

gcc --version
gcc (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Here's the stack trace:

```
(gdb) bt
#0  0xb7fc9559 in __kernel_vsyscall ()
#1  0xb428a2e7 in __pthread_kill_implementation (threadid=threadid@entry=3085974976, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:43
#2  0xb428a35b in __pthread_kill_internal (signo=6, threadid=3085974976) at ./nptl/pthread_kill.c:78
#3  0xb4239111 in __GI_raise (sig=6) at ../sysdeps/posix/raise.c:26
#4  0xb422226a in __GI_abort () at ./stdlib/abort.c:79
#5  0xb467c9d6 in ?? () from /lib/i386-linux-gnu/libstdc++.so.6
#6  0xb468a144 in ?? () from /lib/i386-linux-gnu/libstdc++.so.6
#7  0xb468a1bd in std::terminate() () from /lib/i386-linux-gnu/libstdc++.so.6
#8  0xb468a4bc in __cxa_throw () from /lib/i386-linux-gnu/libstdc++.so.6
#9  0xb5f52543 in cvc5::(anonymous namespace)::CVC5ApiExceptionStream::~CVC5ApiExceptionStream (this=0xbfffef40, __in_chrg=<optimized out>) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:954
#10 0xb5f7a296 in cvc5::Term::getInt64Value (this=0xbffff100) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:3058
#11 0x0040572e in validate_getters () at /home/avj/cvc5/test/api/cpp/sep_log_api.cpp:250
#12 0x0040633d in main () at /home/avj/cvc5/test/api/cpp/sep_log_api.cpp:300
```

If I look inside of Frame 10:

```
(gdb) f 10
#10 0xb5f7a296 in cvc5::Term::getInt64Value (this=0xbffff100) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:3058
3058      CVC5_API_ARG_CHECK_EXPECTED(detail::isInt64(*d_node), *d_node)
```

and then look at what `d_node` is:

```
(gdb) p detail::isInt64(*d_node)
$1 = false
(gdb) p detail::isInt32(*d_node)
$2 = true
```

when doing the `detail::isInt64` check, we eventually get here:

```
(gdb) bt
#0  cvc5::internal::Integer::operator<= (this=0xbffff11c, y=...) at /home/avj/cvc5/src/util/integer_cln_imp.cpp:79
#1  0xb6009041 in cvc5::detail::checkIntegerBounds<long long> (i=...) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:2939
#2  0xb5f77cca in cvc5::detail::isInt64 (node=...) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:2982
#3  0xb5f7a13f in cvc5::Term::getInt64Value (this=0xbffff3c0) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:3058
#4  0x004022c4 in main () at /home/avj/cvc5/test/api/cpp/sep_log_api.cpp:44
```

and if we poke around:

```
cvc5::internal::Integer::operator<= (this=0xbffff11c, y=...) at /home/avj/cvc5/src/util/integer_cln_imp.cpp:79
79        return d_value <= y.d_value;
(gdb) p d_value
$3 = {<cln::cl_RA> = {<cln::cl_R> = {<cln::cl_N> = {<cln::cl_number> = {<cln::cl_gcobject> = {{pointer = 0x1, heappointer = 0x1,
              word = 1}}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}
(gdb) p y.d_value
$4 = {<cln::cl_RA> = {<cln::cl_R> = {<cln::cl_N> = {<cln::cl_number> = {<cln::cl_gcobject> = {{pointer = 0xfffffffd, heappointer = 0xfffffffd,
              word = 4294967293}}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}
(gdb) p d_value <= y.d_value
$5 = false
```

Now, as long as I'm understanding the CLN data structures correctly (http://www.manpagez.com/info/cln/cln-1.3.3/cln_68.php), this seems quite confusing: if `d_value.word` is `1` and `y.d_value.word` is `4294967293`, then surely `d_value <= y.d_value` is `true`, but it returns `false`!

The reason I doubt myself here is that `4294967293` != `std::numeric_limits<int64_t>::max()` -- but the bug is definitely in `detail::isInt64(*d_node)` returning `false`, while `detail::isInt32(*d_node)` returns `true`.

This reproduces the problem:

```cpp
#include <cvc5/cvc5.h>

using namespace cvc5;

int main(void)
{
  Solver slv;

  slv.setLogic("QF_ALL");
  slv.setOption("produce-models", "true");

  Sort integer = slv.getIntegerSort();

  Term p1 = slv.mkConst(integer, "p1");

  Result r(slv.checkSat());

  Term val_for_p1 = slv.getValue(p1);

  val_for_p1.getInt64Value();

  return 0;
}
```

which gives:

```
terminate called after throwing an instance of 'cvc5::CVC5ApiException'
  what():  Invalid argument '0' for '*d_node', expected Term to be a 64-bit integer value when calling getInt64Value()
Aborted
```

Some extra info:

* I worked out how to do this following this log: https://buildd.debian.org/status/fetch.php?pkg=cvc5&arch=i386&ver=1.0.8-1&stamp=1705809275&raw=0
* Due to https://github.com/cvc5/cvc5/pull/10330 I used the same version of cvc5 that that build log uses (1.0.8)
* This build **does not** use `./configure.sh` or `--auto-download`, so here is the version of `cln`: `libcln-dev i386 1.3.6-4+b1`

Here's the bottom of the CMake configure phase:

```
cvc5 1.0.8

Build profile             : debug

Assertions                : on
Debug symbols             : on
Debug context mem mgr     : off

Muzzle                    : off
Statistics                : on
Tracing                   : on

ASan                      : off
UBSan                     : off
TSan                      : off
Coverage (gcov)           : off
Profiling (gprof)         : off
Unit tests                : on
Valgrind                  : off

Shared build              : on
Python bindings           : off
Java bindings             : off
Interprocedural opt.      : off

CryptoMiniSat             : on (system)
GLPK                      : off
Kissat                    : off
LibPoly                   : off
CoCoALib                  : off
MP library                : cln (system)
Editline                  : off

Api docs                  : off
```

**Comment by aytey** [comment]
Okay, so I built using `./configure.sh --auto-download` and got `cln` to build (it failed the configure step).

Here's where it falls apart:

```
#0  cln::compare (x=..., y=...) at /home/avj/cvc5/build/deps/src/CLN-EP/src/integer/elem/cl_I_compare.cc:45
#1  0xb73b8568 in cln::operator<= (x=..., y=...) at /home/avj/cvc5/build/deps/include/cln/integer.h:307
#2  0xb73d5201 in cvc5::internal::Integer::operator<= (this=0xbffff11c, y=...) at /home/avj/cvc5/src/util/integer_cln_imp.cpp:79
#3  0xb6056eb1 in cvc5::detail::checkIntegerBounds<long long> (i=...) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:2939
#4  0xb5fc5bba in cvc5::detail::isInt64 (node=...) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:2982
#5  0xb5fc802f in cvc5::Term::getInt64Value (this=0xbffff3c0) at /home/avj/cvc5/src/api/cpp/cvc5.cpp:3058
#6  0x004022e4 in main () at /home/avj/cvc5/test/api/cpp/sep_log_api.cpp:44
```

```
(gdb) p x
$27 = (const cln::cl_I &) @0xbffff11c: {<cln::cl_RA> = {<cln::cl_R> = {<cln::cl_N> = {<cln::cl_number> = {<cln::cl_gcobject> = {{pointer = 0x1, heappointer = 0x1,
              word = 1}}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}
(gdb) p y
$28 = (const cln::cl_I &) @0xbffff0cc: {<cln::cl_RA> = {<cln::cl_R> = {<cln::cl_N> = {<cln::cl_number> = {<cln::cl_gcobject> = {{pointer = 0xfffffffd, heappointer = 0xfffffffd,
              word = 4294967293}}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}
```

Then we're running this code:

```
(gdb) list
40              // x Fixnum
41              if (fixnump(y))
42                // x Fixnum, y Fixnum
43                { // This assumes cl_value_shift + cl_value_len == cl_pointer_size.
44                  if ((cl_sint)x.word == (cl_sint)y.word) return signean_null;
45                  else if ((cl_sint)x.word > (cl_sint)y.word) return signean_plus;
46                  else return signean_minus;
47                }
48                else
49                // x Fixnum, y Bignum
```

We hit Line 45, and then we have:

```
(gdb) p (cl_sint)y.word
$29 = -3
```

so then CLN reasons that, no, `1` **is not** less than `-3` so it fails the check.

I think the root cause of this is when `9223372036854775807` from `std::numeric_limits<T>::max()` gets packed into a `long` inside of `cvc5::internal::Integer::Integer` (`in64_t`, the param type, is `long long`, while static cast is to `long`, which causes loss of precision).

If you look at in the debugger, you see this:

```
#0  cln::cl_I::cl_I (this=0xbffff0cc, wert=-1) at /home/avj/cvc5/build/deps/include/cln/integer_class.h:60
#1  0xb604baf3 in cvc5::internal::Integer::Integer (this=0xbffff0cc, z=9223372036854775807) at /home/avj/cvc5/src/./util/integer_cln_imp.h:77
```

Notice: `cvc5::internal::Integer::Integer` has a param of `9223372036854775807` but `cln::cl_I::cl_I` has a value of `-1`. Eventually, `cln::cl_I_constructor_from_L` changes `-1` into `0xfffffffd`, which is where our `-3` comes from.

Basically, the bug is here: https://github.com/cvc5/cvc5/blob/71c04db7a45a31e5dd823c5aebbe06c81f958325/src/util/integer_cln_imp.h#L77 and is all because `int64_t` doesn't fit into `long`.

**Comment by swt2c** [comment]
Thanks a lot, @aytey for digging into this issue!

**Comment by aytey** [comment]
@swt2c would it be possible for you to check `main` now to see if this failure is resolved? @mpreiner created a 32-bit build-bot configuration that uses CLN, and it seems that the sep_log_api test is now passing there (and for me locally), so would be good to confirm it is fixed for you too!

**Comment by swt2c** [comment]
Yes, I did just do a quick test on main and it did indeed pass.  I wasn't planning to update the Debian package yet until the next official release of cvc5 though.

Thanks a lot to you both!

**Comment by aytey** [comment]
> I wasn't planning to update the Debian package yet until the next official release of cvc5 though.

No stress from my side, just happy the tests now all pass 🚀

**Comment by swt2c** [comment]
After updating to 1.1.2 (and sorting through the mips64el issues which were related to linking with gold), cvc5 now builds on all primary Debian architectures:
https://buildd.debian.org/status/package.php?p=cvc5&suite=sid

:-)
